File: py/src/apptheory/ids.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Protocol
from uuid import uuid4


class IdGenerator(Protocol):
    def new_id(self) -> str: ...


@dataclass(slots=True)
class RealIdGenerator:
    def new_id(self) -> str:
        return str(uuid4())


@dataclass(slots=True)
class ManualIdGenerator:
    prefix: str
    next: int
    queue: list[str]

    def __init__(self, *, prefix: str = "test-id", start: int = 1) -> None:
        self.prefix = str(prefix)
        self.next = int(start)
        self.queue = []

    def push(self, *ids: str) -> None:
        self.queue.extend([str(v) for v in ids])

    def reset(self) -> None:
        self.next = 1
        self.queue = []

    def new_id(self) -> str:
        if self.queue:
            return self.queue.pop(0)
        out = f"{self.prefix}-{self.next}"
        self.next += 1
        return out

File: ts/src/ids.ts
import { randomUUID } from "node:crypto";

export interface IdGenerator {
  newId(): string;
}

export type IDGenerator = IdGenerator;

export class RandomIdGenerator implements IdGenerator {
  newId(): string {
    return randomUUID();
  }
}

export class ManualIdGenerator implements IdGenerator {
  private _prefix: string;
  private _next: number;
  private _queue: string[];

  constructor(options: { prefix?: string; start?: number } = {}) {
    this._prefix = String(options.prefix ?? "test-id");
    this._next = Number(options.start) || 1;
    this._queue = [];
  }

  queue(...ids: string[]): void {
    this._queue.push(...ids.map((v) => String(v)));
  }

  reset(): void {
    this._next = 1;
    this._queue = [];
  }

  newId(): string {
    if (this._queue.length > 0) {
      return this._queue.shift() ?? "";
    }
    const out = `${this._prefix}-${this._next}`;
    this._next += 1;
    return out;
  }
}
